core/internal/filetransfer/file_transfer_gcs.go

```diff
@@ -54,5 +77,228 @@ func (ft *GCSFileTransfer) Upload(task *ReferenceArtifactUploadTask) error {
+		g := new(errgroup.Group)
+		g.SetLimit(maxWorkers)
+		for _, name := range objectNames {
+			objName := name // for closure in the goroutine
+			g.Go(func() error {
+				object, _, err := ft.getObjectAndAttrs(bucket, task, objName)
+				if err != nil {
+					if errors.Is(err, ErrObjectIsDirectory) {
+						ft.logger.Debug(
+							"GCSFileTransfer: Download: skipping reference because it seems to be a folder",
+							"reference", reference,
+							"object", objName,
+						)
+						return nil
+					}
+					return formatDownloadError(fmt.Sprintf("error getting object %s", objName), err)
+				}
+				localPath := getDownloadFilePath(objName, objectName, task.Path)
+				return ft.downloadFile(object, localPath)
+			})
+		}
+
+		// Wait for all of the go routines to complete and return an error if any errored out
+		if err := g.Wait(); err != nil {
+			return formatDownloadError("", err)
+		}
+	} else {
+		object, objAttrs, err := ft.getObjectAndAttrs(bucket, task, objectName)
+		if err != nil {
+			if errors.Is(err, ErrObjectIsDirectory) {
+				ft.logger.Debug(
+					"GCSFileTransfer: Download: skipping reference because it seems to be a folder",
+					"reference", reference,
+				)
+				return nil
+			}
+			return formatDownloadError(fmt.Sprintf("error getting object %s", objectName), err)
+		}
+		if objAttrs.Etag != task.Digest {
+			err := fmt.Errorf("digest/etag mismatch: etag %s does not match expected digest %s", objAttrs.Etag, task.Digest)
+			return formatDownloadError("", err)
+		}
+		return ft.downloadFile(object, task.Path)
+	}
+	return nil
+}
+
+// DownloadFile downloads the contents of object into a file at path localPath
+func (ft *GCSFileTransfer) downloadFile(object *storage.ObjectHandle, localPath string) error {
+	r, err := object.NewReader(ft.ctx)
+	if err != nil {
+		return formatDownloadError("error creating reader", err)
+	}
+	defer func(file io.ReadCloser) {
+		if err := file.Close(); err != nil {
+			ft.logger.CaptureError(
+				fmt.Errorf(
+					"file transfer: download: error closing response reader: %v",
+					err,
+				))
+		}
+	}(r)
+
+	// Check if the directory exists, and create it if it doesn't
+	dir := path.Dir(localPath)
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		if err := os.MkdirAll(dir, os.ModePerm); err != nil {
+			return formatDownloadError(
+				fmt.Sprintf("error creating download directory %s", dir),
+				err,
+			)
+		}
+	} else if err != nil {
+		return formatDownloadError(
+			fmt.Sprintf("error finding download directory %s", dir),
+			err,
+		)
+	}
+
+	// open the file for writing and defer closing it
+	file, err := os.Create(localPath)
+	if err != nil {
+		return formatDownloadError(
+			fmt.Sprintf("error creating download file %s", localPath),
+			err,
+		)
+	}
+	defer func(file *os.File) {
+		if err := file.Close(); err != nil {
+			ft.logger.CaptureError(
+				formatDownloadError(
+					fmt.Sprintf("error closing file %s", localPath),
+					err,
+				))
+		}
```

Comment:
In this case, if `Close()` returns an error, then the method can't guarantee it fulfilled its contract (of downloading the entire contents of the file), so the method should return the error rather than just logging it. You can use a named return value for this, like here: https://github.com/wandb/wandb/blob/32246d756dc9cd3ea5f04918a4726b37d6cc08fd/core/internal/runconsolelogs/linefile.go#L64-L71